### ISSUE TITLE
#error error in arpa/nameser_compat.h

### ISSUE THREAD

**Issue by stefanct** [omniti-labs/illumos-omnios#7]
There is a `#` sign missing in [`arpa/nameser_compat.h` at line 121](https://github.com/omniti-labs/illumos-omnios/blob/c2e7b48d563d0e56b74d853118918af352e75cbb/usr/src/head/arpa/nameser_compat.h#L121) (and the trailing semicolon is superfluous). Why I did hit this with a virtual x86 machine is another question I am still investigating :)

**Comment by danmcd** [comment]
A damned good question.  It's a bug upstream too, as there's no difference between our version and illumos-gate's:

`bloody(~/ws)[0]% diff illumos-{gate,omnios}/usr/src/head/arpa/nameser_compat.h
bloody(~/ws)[0]% 
`

**Comment by stefanct** [comment]
I figured that much but I dont have nor want an account there to report a bug ;) So by all means, please report it!
Also, I wanted to find out why I end up there at all.

Some context: I was building a dependency for my own project from source, namely pciutils/libpci (https://git.kernel.org/cgit/utils/pciutils/pciutils.git/). While that supports SunOS in general I could only get some very old versions (~2008 or something) of it to actually compile. The following patch fixes the problem for the current version... it simply removes some SunOS-specific workarounds regarding endianess... however, I have no idea if this should be upstreamed because it is no longer necessary in any Solaris-based system. Any idea why that ever was needed and if I should propose inclusion of the change to upstream?

	diff --git a/lib/sysdep.h b/lib/sysdep.h
	index b864e90..5c6b91d 100644
	--- a/lib/sysdep.h
	+++ b/lib/sysdep.h
	@@ -38,24 +38,6 @@ typedef u16 word;
	 #define BIG_ENDIAN __BIG_ENDIAN
	 #endif
	 
	-#ifdef PCI_OS_SUNOS
	-#include <sys/byteorder.h>
	-#if defined(__i386) && defined(LITTLE_ENDIAN)
	-# define BYTE_ORDER LITTLE_ENDIAN
	-#elif defined(__sparc) && defined(BIG_ENDIAN)
	-# define BYTE_ORDER BIG_ENDIAN
	-#else
	-#define BIG_ENDIAN 4321
	-#endif
	-#ifndef BYTE_ORDER
	-#ifdef _LITTLE_ENDIAN
	-#define BYTE_ORDER 1234
	-#else
	-#define BYTE_ORDER 4321
	-#endif
	-#endif /* BYTE_ORDER */
	-#endif /* PCI_OS_SUNOS */
	-
	 #ifdef PCI_OS_WINDOWS
	 #ifdef __MINGW32__
	   #include <sys/param.h>

**Comment by danmcd** [comment]
Odd... we build PCIutils on with omnios-build, and we apparently use the command-line (changes happened before I got here):

https://github.com/omniti-labs/omnios-build/blame/master/build/pciutils/build.sh

**Comment by trisk** [comment]
The `error` line probably predates the standardisation of the `#error` preprocessor directive. I don't know how you could end up with neither `__amd64__` nor `__i386` defined, as both gcc and Studio use those macros.

`arpa/nameser_compat.h` is probably pulled in via `<resolv.h>` or similar, so this is an accidental leakage and I wouldn't count on it existing in the future.

The `BYTE_ORDER` macros are originally a BSDism, I believe, but part of a recent proposal for POSIX `<endian.h>`: http://austingroupbugs.net/view.php?id=162

We should probably adopt some form of this.

**Comment by stefanct** [comment]
@danmcd well, uh, maybe I am missing something, but that script explicitly sets `BYTE_ORDER=1234` and `LITTLE_ENDIAN=1234` for the build? :)

@trisk: when we get to that header BYTE_ORDER is already set by the bogus pciutils `sysdep.h` but `LITTLE_ENDIAN` is not!
Hence the expanded branch condition is: ` 1234 != 4321 && 1234 != LITTLE_ENDIAN && 1234 != PDP_ENDIAN`

**Comment by danmcd** [comment]
Because OmniOS doesn't officially support SPARC.  As for endian.h, yes, we should probably support that.

**Comment by stefanct** [comment]
Yes, but it also overrides the broken autodetection inside the pciutils code... so both, Solaris distros and pciutils jump through hoops completely unnecessarily and break simple compilation from source - meh.

**Comment by danmcd** [comment]
So we fix it in illumos, and there's already a longstanding bug open for it:

https://illumos.org/issues/2169

it wasn't fixed, BTW, because of a desire to just do endian.h per @trisk.  I believe we should fix 2169, and will ping the list shortly.
